core-common/src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java

```diff
@@ -313,33 +314,59 @@ public void syncPostBatch(
                 item.setBrokerId(partitionToBroker.get(item.getPartition()));
             });
 
-            final Map<BatchItem, CompletableFuture<Exception>> sendFutures = new HashMap<>();
+            final List<CompletableFuture<Exception>> futures = new LinkedList<>();
+            final Map<String, CompletableFuture<Exception>> futuresByKey = new HashMap<>();
+
             for (final BatchItem item : batch) {
-                final String brokerId = item.getBrokerId();
-                if (brokerId == null) {
+                if (item.getBrokerId() == null) {
                     item.updateStatusAndDetail(EventPublishingStatus.FAILED,
                             String.format("No leader for partition: %s, topic: %s.", item.getPartition(), topicId));
                     LOG.error("Failed to publish to kafka. No leader for ({}:{}).",
                             topicId, item.getPartition());
                     continue;
                 }
-                item.setStep(EventPublishingStep.PUBLISHING);
-                sendFutures.put(item, publishItem(producer, topicId, eventType, item, delete));
+
+                final String itemKey = item.getEventKey();
+                if (itemKey == null) {
+                    // items without a key go out of order
+                    futures.add(publishItem(producer, topicId, eventType, item, delete));
+                } else {
+                    // items for a single key are composed together for sequential publishing
+                    CompletableFuture<Exception> fut = futuresByKey.get(itemKey);
+                    if (fut == null) {
+                        fut = publishItem(producer, topicId, eventType, item, delete);
+                    } else {
+                        fut = fut.thenCompose((e) -> {
```

Comment:
Maybe we want to use `thenComposeAsync()` here, as `publishItem()` may block on the next `producer.send()` call...